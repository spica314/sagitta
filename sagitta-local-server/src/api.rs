use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;

use crate::api_state::ApiState;

use self::v1::sync::v1_sync;

pub mod v1;

pub struct ServerConfig {
    pub clock: Clock,
    pub port: u16,
    pub local_system_workspace_base_path: PathBuf,
    pub remote_api_base_url: String,
}

pub async fn run_local_api_server(config: ServerConfig) {
    let state = ApiState::new(
        config.clock.clone(),
        config.local_system_workspace_base_path.clone(),
        &config.remote_api_base_url,
    )
    .await;

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(v1_sync)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
