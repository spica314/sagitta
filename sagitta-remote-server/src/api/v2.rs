pub mod commit;
pub mod create_workspace;
pub mod get_attr;
pub mod get_file_blob_id;
pub mod get_workspace_id_from_name;
pub mod get_workspaces;
pub mod read_blob;
pub mod read_dir;
pub mod sync_files_with_workspace;
pub mod write_blob;
