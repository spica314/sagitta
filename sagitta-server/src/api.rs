use std::path::PathBuf;

use actix_web::{web, App, HttpServer};
use sagitta_common::clock::Clock;
use sagitta_objects::{SagittaTreeObject, SagittaTreeObjectFile};

use crate::state::ApiState;

use self::{
    blob::read::blob_read,
    trunk::get_head::trunk_get_head,
    workspace::{create::workspace_create, get_head::workspace_get_head, list::workspace_list},
};

pub mod blob;
pub mod trunk;
pub mod workspace;

pub struct ServerConfig {
    pub base_path: PathBuf,
    pub surreal_uri: String,
    pub is_main: bool,
    pub clock: Clock,
    pub port: u16,
}

pub async fn run_server(config: ServerConfig) {
    let state = ApiState::new(
        config.base_path.clone(),
        config.clock.clone(),
        &config.surreal_uri,
        config.is_main,
    )
    .await;

    // root (dir1)
    // - hello.txt (file1)
    // - hello_dir (dir2)
    //     - hello2.txt (file2)

    let file1_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!\n".as_slice())
        .unwrap();
    let file1 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file1_blob_id,
        size: 14,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file1)
        .unwrap();

    let file2_blob_id = state
        .remote_system_workspace_manager
        .save_object(None, b"Hello, world!!\n".as_slice())
        .unwrap();
    let file2 = SagittaTreeObject::File(SagittaTreeObjectFile {
        blob_id: file2_blob_id,
        size: 15,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o644,
    });
    let file2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &file2)
        .unwrap();

    let tree2 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![("hello2.txt".to_string(), file2_id)],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree2_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree2)
        .unwrap();

    let tree1 = SagittaTreeObject::Dir(sagitta_objects::SagittaTreeObjectDir {
        items: vec![
            ("hello.txt".to_string(), file1_id),
            ("hello_dir".to_string(), tree2_id),
        ],
        size: 4096,
        mtime: config.clock.now(),
        ctime: config.clock.now(),
        perm: 0o755,
    });
    let tree1_id = state
        .remote_system_workspace_manager
        .save_tree(None, &tree1)
        .unwrap();

    let commit = sagitta_objects::SagittaCommitObject {
        tree_id: tree1_id,
        parent_commit_id: None,
        message: "Initial commit".to_string(),
    };
    let commit_id = state
        .remote_system_workspace_manager
        .save_commit(None, &commit)
        .unwrap();
    state
        .remote_system_workspace_manager
        .set_trunk_head(&commit_id)
        .await
        .unwrap();

    HttpServer::new(move || {
        App::new()
            .app_data(web::Data::new(state.clone()))
            .service(blob_read)
            .service(trunk_get_head)
            .service(workspace_create)
            .service(workspace_list)
            .service(workspace_get_head)
    })
    .bind(("0.0.0.0", config.port))
    .unwrap()
    .run()
    .await
    .unwrap();
}
