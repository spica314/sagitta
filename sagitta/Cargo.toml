[package]
name = "sagitta"
version = "0.1.0"
edition = "2021"

[dependencies]
actix-web = { workspace = true }
env_logger = { workspace = true }
fuser = { workspace = true }
insta = { workspace = true }
libc = { workspace = true }
log = { workspace = true }
serde = { workspace = true }
serde_cbor = { workspace = true }
serde_json = { workspace = true }
tempfile = { workspace = true }
tokio = { workspace = true }
ureq = { workspace = true }

sagitta-objects = { workspace = true }
sagitta-api-schema = { workspace = true }
sagitta-objects-store = { workspace = true }
