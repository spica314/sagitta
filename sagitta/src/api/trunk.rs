pub mod get_head;
