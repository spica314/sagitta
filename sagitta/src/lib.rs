pub mod api_client;
pub mod args;
pub mod fs;
