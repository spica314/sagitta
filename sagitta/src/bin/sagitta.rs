use sagitta::fs::{run_fs, SagittaConfig};

fn main() {
    env_logger::init();
    let uid = unsafe { libc::getuid() };
    let gid = unsafe { libc::getgid() };
    let config = SagittaConfig {
        base_url: "http://localhost:8081".to_string(),
        mountpoint: "./sagitta-test".to_string(),
        uid,
        gid,
    };
    run_fs(config);
}
