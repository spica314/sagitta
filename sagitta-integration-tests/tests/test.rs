use std::{process::Command, time::SystemTime};

use sagitta::{
    api_client::SagittaApiClient,
    fs::{run_fs, SagittaConfig},
};
use sagitta_objects::SagittaTreeObject;
use sagitta_server::api::ServerConfig;
use serial_test::serial;
use tempfile::tempdir;
use tokio::runtime::Builder;

#[test]
#[serial]
fn test_1() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let port = 8081;
    let config = ServerConfig {
        base_path: tempdir.as_ref().to_path_buf(),
        clock: sagitta_server::tools::Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let client = SagittaApiClient::new(format!("http://localhost:{}", port));

    let head_id_res = client.trunk_get_head().unwrap();
    let head_id = head_id_res.id;
    insta::assert_debug_snapshot!(head_id);

    let commit = client.blob_read_as_commit_object(&head_id).unwrap();
    insta::assert_debug_snapshot!(commit);

    let dir = client.blob_read_as_tree_object(&commit.tree_id).unwrap();
    insta::assert_debug_snapshot!(dir);

    let SagittaTreeObject::Dir(dir) = dir else {
        panic!()
    };
    let mut xs = vec![];
    for item in &dir.items {
        let file = client.blob_read_as_tree_object(&item.1).unwrap();
        let SagittaTreeObject::File(file) = file else {
            continue;
        };

        let blob_res = client.blob_read(&file.blob_id).unwrap();
        let blob = std::str::from_utf8(blob_res.blob.as_slice()).unwrap();
        insta::assert_debug_snapshot!(blob);

        xs.push((item.0.clone(), blob.to_string()));
    }
    insta::assert_debug_snapshot!(xs);
}

#[test]
#[serial]
fn test_2() {
    let runtime = Builder::new_multi_thread()
        .worker_threads(1)
        .enable_all()
        .build()
        .unwrap();

    let tempdir1 = tempdir().unwrap();
    let fixed_system_time =
        SystemTime::UNIX_EPOCH + std::time::Duration::from_secs(40 * 365 * 24 * 60 * 60);
    let port = 8082;
    let config = ServerConfig {
        base_path: tempdir1.as_ref().to_path_buf(),
        clock: sagitta_server::tools::Clock::new_with_fixed_time(fixed_system_time),
        port,
    };

    runtime.spawn(async {
        sagitta_server::api::run_server(config).await;
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let tempdir2 = tempdir().unwrap();
    let tempdir2_str = tempdir2.as_ref().to_str().unwrap().to_string();
    std::thread::spawn(move || {
        let config = SagittaConfig {
            base_url: format!("http://localhost:{}", port),
            mountpoint: tempdir2_str,
            uid: 0,
            gid: 0,
        };
        run_fs(config);
    });
    std::thread::sleep(std::time::Duration::from_secs(1));

    let out1 = Command::new("ls")
        .arg("-lA")
        .current_dir(tempdir2.path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out1);

    let out2 = Command::new("cat")
        .arg("hello.txt")
        .current_dir(tempdir2.path())
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out2);

    let mut out3_dir = tempdir2.path().to_path_buf();
    out3_dir.push("hello_dir");
    let out3 = Command::new("ls")
        .arg("-lA")
        .current_dir(&out3_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out3);

    let out4 = Command::new("cat")
        .arg("hello2.txt")
        .current_dir(out3_dir)
        .output()
        .expect("failed to execute process");
    insta::assert_debug_snapshot!(out4);
}
