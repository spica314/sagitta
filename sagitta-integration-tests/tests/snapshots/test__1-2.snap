---
source: sagitta-integration-tests/tests/test.rs
expression: commit
---
SagittaCommitObject {
    tree_id: ObjectId {
        id: "38dc07c664c7ebafbc6c93b9454f47ec9c1a3f20cac908edd23e08f35bf2fc1f",
    },
    parent_commit_id: None,
    message: "Initial commit",
}
